rpc/lib/client/http_client.go

```diff
@@ -181,27 +295,83 @@ func (c *URIClient) SetCodec(cdc *amino.Codec) {
 
 //------------------------------------------------
 
-func unmarshalResponseBytes(cdc *amino.Codec, responseBytes []byte, result interface{}) (interface{}, error) {
+func unmarshalResponseBytes(cdc *amino.Codec, responseBytes []byte, expectedID types.JSONRPCStringID, result interface{}) (interface{}, error) {
 	// Read response.  If rpc/core/types is imported, the result will unmarshal
 	// into the correct type.
 	// log.Notice("response", "response", string(responseBytes))
 	var err error
 	response := &types.RPCResponse{}
 	err = json.Unmarshal(responseBytes, response)
 	if err != nil {
-		return nil, errors.Errorf("Error unmarshalling rpc response: %v", err)
+		return nil, errors.Errorf("error unmarshalling rpc response: %v", err)
 	}
 	if response.Error != nil {
-		return nil, errors.Errorf("Response error: %v", response.Error)
+		return nil, errors.Errorf("response error: %v", response.Error)
+	}
+	// From the JSON-RPC 2.0 spec:
+	//  id: It MUST be the same as the value of the id member in the Request Object.
+	if err := validateResponseID(response, expectedID); err != nil {
+		return nil, err
 	}
 	// Unmarshal the RawMessage into the result.
 	err = cdc.UnmarshalJSON(response.Result, result)
 	if err != nil {
-		return nil, errors.Errorf("Error unmarshalling rpc response result: %v", err)
+		return nil, errors.Errorf("error unmarshalling rpc response result: %v", err)
 	}
 	return result, nil
 }
 
+func unmarshalResponseBytesArray(cdc *amino.Codec, responseBytes []byte, expectedID types.JSONRPCStringID, results []interface{}) ([]interface{}, error) {
+	var (
+		err       error
+		responses []types.RPCResponse
+	)
+	err = json.Unmarshal(responseBytes, &responses)
+	if err != nil {
+		return nil, errors.Errorf("error unmarshalling rpc response: %v", err)
+	}
+	// No response error checking here as there may be a mixture of successful
+	// and unsuccessful responses.
+
+	if len(results) != len(responses) {
+		return nil, errors.Errorf("expected %d result objects into which to inject responses, but got %d", len(responses), len(results))
+	}
+
+	// From the JSON-RPC 2.0 spec:
+	//  id: It MUST be the same as the value of the id member in the Request Object.
+	for i, response := range responses {
+		if err := validateResponseID(&response, expectedID); err != nil {
+			return nil, errors.Errorf("failed to validate response ID in response %d: %v", i, err)
+		}
+	}
+
+	// Unmarshal the Result fields into the final result objects
+	for i := 0; i < len(responses); i++ {
```

Comment:
same question. why not `range`?

also can't we merge this and the above loop?

Reply:
As above.

I could've merged the loop with the one above it, but I was doing all the error checking first for ID validation, and if all responses pass that check then go through the effort of unmarshalling.